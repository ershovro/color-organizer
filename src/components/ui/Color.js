import React from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router'
import StarRating from './StarRating'
import TimeAgo from './TimeAgo'
import {FaTrash} from 'react-icons/fa'
import '../../stylesheets/Color.scss'

class Color extends React.Component {

   render() {
      const { id, title, color, rating, timeStamp, onRemove, onRate, history } = this.props;

      return (
          <section className="color" style={this.style}>
             <h1
                ref="title"
                className="color__title"
                onClick={() => history.push(`/${id}`)}>
                {title}
             </h1>
             <button 
                className="color__removeButton"
                onClick={onRemove}>
                <FaTrash/>
             </button>
             <div className="color__timestamp">
                <TimeAgo timeStamp={timeStamp}/>
             </div>
             <div
                className="color__canvas"
                onClick={() => history.push(`/${id}`)}
                style={{ backgroundColor: color }}>
             </div>
             <div className="color__starRating">
                 <StarRating starsSelected={rating} onRate={onRate}/>
             </div>
          </section>
      );
   }

}

Color.propTypes = {
    title: PropTypes.string.isRequired,
    color: PropTypes.string.isRequired,
    rating: PropTypes.number,
    onRemove: PropTypes.func,
    onRate: PropTypes.func
}

Color.defaultProps = {
    rating: 0,
    onRemove: f => f,
    onRate: f => f
}

export default withRouter(Color);
