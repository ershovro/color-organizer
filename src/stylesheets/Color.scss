@import "globals";

.color {
  @include shadow;
  
  height: 235px;
  
  border: 2px solid $light-grey;
  border-bottom-color: darken($light-grey, 40%);
  border-right-color: darken($light-grey, 40%);
  
  margin: 1em;
  padding: 0;
  
 
  display: inline-flex;
  flex-direction: column;
  
  position: relative;
}

.color__removeButton {
   position: absolute;
   top: .5em;
   right: .5em;
   
   
}

.color__timestamp {
  position: absolute;
  bottom: .5em;
  right: .5em;
  font-family: monospace;
  color: darken($light-grey, 40%);
}

.color__canvas {
  height: 100px;
}