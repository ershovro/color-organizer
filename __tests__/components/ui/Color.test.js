import React from 'react';
import Color from "../../../src/components/ui/Color";
import enzyme from 'enzyme';
import { JSDOM } from 'jsdom';
import toJSON from 'enzyme-to-json'
import { compose } from 'redux';
import Adapter from 'enzyme-adapter-react-16';


const { window } = new JSDOM(`...`);
const { document } = (new JSDOM(`...`)).window;

enzyme.configure({ adapter: new Adapter() });
global.window = window;
global.document = document;

describe('<Color /> ui component', () => {

   describe('Rendering ui', () => {

      it('Renders Correctly', () => {
         compose(expect, toJSON, enzyme.shallow)(
            <Color
               title="god color"
               color="#F0F0F0"
               rating={5}
               timeStamp="Mon Apr 11 2016 12:54:19 GMT-0700 (PDT)"
            />
         ).toMatchSnapshot();
      });

      it('Renders section with class .color', () => {
         expect(
            enzyme.mount( <Color title="god color" color="#F2F5F4"/> ).
            find('section.color').
            first().length
         ).toBe(1)
      });

      it('Clicking default remove button do not cause error', () => {
         enzyme.mount( <Color title="god color" color="#F2F5F4"/> ).
         find('button.color__removeButton').
         first().
         simulate('click')
      });

      it('Clicking default rate button do not cause error', () => {
         enzyme.mount( <Color title="god color" color="#F2F5F4"/> ).
         find('div.star').
         first().
         simulate('click')
      });
   });

   describe('Removing color', () => {
      let wrapper;
      let _remove = jest.fn();

      beforeEach(() => {
         wrapper = enzyme.mount( <Color onRemove={_remove} title="god color" color="#F2F5F4"/> );
      });
      afterEach(() => {
         _remove.mockReset();
      });

      it('invoke onRemove handler', () => {
         wrapper.
         find('button.color__removeButton').
         first().
         simulate('click');

         expect(_remove).toBeCalled();
      });
   });

   describe('Rating color', () => {
      let wrapper;
      let _rate = jest.fn();

      beforeEach(() => {
         wrapper = enzyme.mount( <Color onRate={_rate} title="god color" color="#F2F5F4"/> );
      });
      afterEach(() => {
         _rate.mockReset();
      });

      it('invoke onRate handler', () => {
         wrapper.
         find('div.star').
         first().
         simulate('click');

         expect(_rate).toBeCalled();
      });

      it('invoke onRate handler with correct arguments', () => {
         wrapper.
         find('div.star').
         first().
         simulate('click');

         expect(_rate).toBeCalledWith(1);
      });
   });
});

