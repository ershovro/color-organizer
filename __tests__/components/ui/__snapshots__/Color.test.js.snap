// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Color /> ui component Rendering ui Renders Correctly 1`] = `
<section
  className="color"
>
  <h1
    className="color__title"
    onClick={[Function]}
  >
    god color
  </h1>
  <button
    className="color__removeButton"
    onClick={[Function]}
  >
    <FaTrash />
  </button>
  <div
    className="color__timestamp"
  >
    <TimeAgo
      timeStamp={2016-04-11T19:54:19.000Z}
    />
  </div>
  <div
    className="color__canvas"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#F0F0F0",
      }
    }
  />
  <div
    className="color__starRating"
  >
    <StarRating
      onRate={[Function]}
      starsSelected={5}
    />
  </div>
</section>
`;
